src/csr.tex

```diff
@@ -157,19 +157,23 @@ \section{CSR Instructions}
 
 \subsection*{CSR Access Ordering}
 
-On a given hart, explicit and implicit CSR access are performed in program
-order with respect to those instructions whose execution behavior is affected
-by the state of the accessed CSR. In particular, a CSR access is performed
+Each RISC-V hart normally observes its own CSR accesses, including its
+implicit CSR accesses, as performed in program order.
+In particular, unless specified otherwise, a CSR access is performed
 after the execution of any prior instructions in program order whose behavior
 modifies or is modified by the CSR state and before the execution of any
 subsequent instructions in program order whose behavior modifies or is modified
-by the CSR state. Furthermore, a CSR read access instruction returns the
-accessed CSR state before the execution of the instruction, while a CSR write
-access instruction updates the accessed CSR state after the execution of the
+by the CSR state.
+Any side effects from a CSR access are likewise observed to occur
+synchronously in program order, unless documented otherwise.
+Furthermore, an explicit CSR read returns the
+CSR state before the execution of the instruction, while an
+explict CSR write updates the CSR state after the execution of the
```

Comment:
> updates the CSR state after the execution

shouldn't that rather be 
>updates the CSR state at the end of the execution

Or words to that effect (the state is after all changed _by_ the instruction execution.

Reply:
Perhaps it would be clearer to say that an explicit CSR write suppresses and overrides any implicit writes or modifications to the same CSR by the same instruction.

Reply:
yes. I agree but perhaps:
> by the same instruction.

to be more inclusive of any concurrent change including side effects of the instruction, with wording:
> during the same instruction,  ...  

?

... , unless documented otherwise.

;-)  

This qualifier could, perhaps, be a prefix to the full text, but I do think it helps to be explicit at each point of potential exception.

In effect these becomes the "unless otherwise specified" exception within the "this is the default behaviour" prefix.  

So, should we document the "default behaviour" behaviour more explicitly than "normally"?

Although we are providing a precise description, and I like it, does the general reader understand the context? Do we need to provide some background in the"default behaviour" prefix?

Something like "CSRs are unique within the RV model as their values can change independent of currently executing code or can have side-effects related to execution of that code., including execution of instructions that explicitly access/modify CSRs"

Reply:
David Horner wrote:
> yes. I agree but perhaps:
> > by the same instruction.
>
> to be more inclusive of any concurrent change including side effects of the instruction, with wording:
> > during the same instruction, ...

The ISA has no requirements for what must happen "during an instruction" except what is caused by the instruction itself.  For that reason, "during the same instruction" wouldn't restrict anything more than the words "by the same instruction" do.

When executing a CSR instruction, the latter wording gives the hardware the freedom to make independent, concurrent reads/writes to the same CSR appear to occur entirely before or after the CSR instruction, or to be suppressed (because the independent accesses occur after the instruction reads the CSR but before the instruction writes the CSR), or some mixture.  Again, because the ISA never says what independent events must occur "during an instruction", all of these possible outcomes seem legitimate to me, at least until there are other rules with more to say (presumably about specific CSRs and specific independent reads/writes).  But as those other rules wouldn't contravene what is written here about suppressing implicit CSR accesses by the same instruction, we also don't need to add here the usual fudge, "unless specified otherwise".